[Navigation] Provide better error reporting

rwlbuis opened the pull request:
#### 9cfb6bc898ed06e74bff228236d2405dbe3f67d9
<pre>
[Navigation] Provide better error reporting
<a href="https://bugs.webkit.org/show_bug.cgi?id=279032">https://bugs.webkit.org/show_bug.cgi?id=279032</a>

Reviewed by Alex Christensen.

Provide better error reporting by refactoring code in SerializedScriptValue that extracts detailed
error information from JSC::ErrorInstance and reuse the new method both for error reporting when
aborting the navigation and on handler promise rejection.

* LayoutTests/http/wpt/navigation-api/transition-promises-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/signal-abort-window-stop-after-intercept-expected.txt:
* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
(WebCore::CloneSerializer::dumpIfTerminal):
(WebCore::extractErrorInformationFromErrorInstance):
* Source/WebCore/bindings/js/SerializedScriptValue.h:
* Source/WebCore/page/Navigation.cpp:
(WebCore::Navigation::abortOngoingNavigation):
(WebCore::Navigation::innerDispatchNavigateEvent):

Canonical link: <a href="https://commits.webkit.org/283197@main">https://commits.webkit.org/283197@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1edf293e1615e0a8cb4d8a0718271d3bc0cef50c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/65442 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/44813 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/18060 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/69468 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/16051 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/67560 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/52614 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/16331 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/52558 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/11129 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/68509 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/41418 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/56646 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/33181 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/38091 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/14021 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/14927 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/59927 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/14362 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/71173 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/9396 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/13825 "Found 1 new test failure: imported/w3c/web-platform-tests/media-source/mediasource-addsourcebuffer-mode.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/59878 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/9428 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/56707 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/60151 "Found 2 new API test failures: /WebKitGTK/TestWebKitWebContext:/webkit/WebKitWebContext/uri-scheme, /WebKitGTK/TestWebKitAccessibility:afterAll (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14445 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/7768 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/1422 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/40623 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/41699 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/42882 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/41443 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

TingPing reviewed (commented):
I like this approach much more than mine

achristensen07 commented on Source/WebCore/bindings/js/SerializedScriptValue.cpp:
```diff
@@ -6477,4 +6439,52 @@ IDBValue SerializedScriptValue::writeBlobsToDiskForIndexedDBSynchronously()
     return value;
 }
 
+bool extractErrorInformationFromErrorInstance(JSC::JSGlobalObject* lexicalGlobalObject, ErrorInstance& errorInstance, String& errorTypeString, String& message, unsigned& line, unsigned& column, String& sourceURL, String& stack)
```
Looks good, but new code should not use out params like this.

struct ErrorInformation {
...
}
std::optional<ErrorInformation> extractErrorInformationFromErrorInstance(JSC::JSGlobalObject*, ...)

rwlbuis replied:
Agreed, I hope I addressed it.

achristensen07 commented on Source/WebCore/bindings/js/SerializedScriptValue.cpp:
```diff
@@ -6477,4 +6434,53 @@ IDBValue SerializedScriptValue::writeBlobsToDiskForIndexedDBSynchronously()
+    }
+    RETURN_IF_EXCEPTION(scope, std::nullopt);
+
+    return errorInformation;
```
This is great.  It would be even better if we didn't use the default constructor then populate the struct, but rather gathered the members then only call a constructor if we have all the pieces.  This is how the modern deserialization code is generated.  That'll make it so we get a compiler failure if we forget one.

Otherwise, I think this needs to be like this to prevent unnecessary copy constructors:
return { WTFMove(errorInformation) };

rwlbuis replied:
Sure, makes sense, I tried to implement that.

achristensen07 commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -859,14 +866,17 @@ bool Navigation::innerDispatchNavigateEvent(NavigationNavigationType navigationT
+                errorMessage = makeString("Uncaught "_s, errorInformation->errorTypeString, ": "_s, errorInformation->message);
+            }
+            if (!errorInformation)
+                errorInformation = { };
```
This doesn't do what you think it does.  It checks if errorInformation is nullopt, then re-assigns it to nullopt if it is, and line 879 will still crash.

rwlbuis replied:
Should be fixed.

achristensen07 commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -633,13 +633,20 @@ void Navigation::abortOngoingNavigation(NavigateEvent& event)
+    if (auto* errorInstance = jsDynamicCast<JSC::ErrorInstance*>(error))
+        errorInformation = extractErrorInformationFromErrorInstance(protectedScriptExecutionContext()->globalObject(), *errorInstance);
+    if (!errorInformation)
+        errorInformation = { };
```
This doesn't do anything.

rwlbuis replied:
Fixed.

achristensen07 commented on Source/WebCore/bindings/js/SerializedScriptValue.cpp:
```diff
@@ -6477,4 +6434,58 @@ IDBValue SerializedScriptValue::writeBlobsToDiskForIndexedDBSynchronously()
+    RETURN_IF_EXCEPTION(scope, std::nullopt);
+
+    ErrorInformation errorInformation = { errorTypeString, message, line, column, sourceURL, stack };
+    return { WTFMove(errorInformation) };
```
This can just be return { ErrorInformation { ... } };

rwlbuis replied:
Done.

achristensen07 commented on Source/WebCore/bindings/js/SerializedScriptValue.h:
```diff
@@ -62,6 +62,10 @@ class MemoryHandle;
 #endif
 
+namespace JSC {
+class ErrorInstance;
```
I'm a little surprised JSGlobalObject doesn't need a forward declaration too.  May as well add it?

rwlbuis replied:
Added.

achristensen07 commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -633,13 +633,20 @@ void Navigation::abortOngoingNavigation(NavigateEvent& event)
+    ErrorInformation errorInformation;
+    if (auto* errorInstance = jsDynamicCast<JSC::ErrorInstance*>(error)) {
+        if (auto result = extractErrorInformationFromErrorInstance(protectedScriptExecutionContext()->globalObject(), *errorInstance))
+            errorInformation = *result;
```
WTFMove(*result);

rwlbuis replied:
Done.

achristensen07 commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -859,14 +866,17 @@ bool Navigation::innerDispatchNavigateEvent(NavigationNavigationType navigationT
-                errorMessage = makeString("Uncaught "_s, error->sanitizedToString(protectedScriptExecutionContext()->globalObject()));
+            if (auto* errorInstance = jsDynamicCast<JSC::ErrorInstance*>(result)) {
+                if (auto result = extractErrorInformationFromErrorInstance(protectedScriptExecutionContext()->globalObject(), *errorInstance)) {
+                    errorInformation = *result;
```
WTFMove(*result);

rwlbuis replied:
Done.

achristensen07 commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -633,13 +633,20 @@ void Navigation::abortOngoingNavigation(NavigateEvent& event)
 
-    // FIXME: Fill in exception information.
-    dispatchEvent(ErrorEvent::create(eventNames().navigateerrorEvent, exception.message(), { }, 0, 0, { globalObject->vm(), domException }));
+    dispatchEvent(ErrorEvent::create(eventNames().navigateerrorEvent, exception.message(), errorInformation.sourceURL, errorInformation.line, errorInformation.column, { globalObject->vm(), domException }));
```
Do we want to do this if there is no error information?

rwlbuis replied:
I think so, but maybe we want to log that case in the future, as it should not happen.

achristensen07 commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -859,14 +866,17 @@ bool Navigation::innerDispatchNavigateEvent(NavigationNavigationType navigationT
 
-            // FIXME: Set line/column numbers.
-            dispatchEvent(ErrorEvent::create(eventNames().navigateerrorEvent, exception.message(), document->url().stringWithoutFragmentIdentifier(), 1, 1, { protectedScriptExecutionContext()->globalObject()->vm(), result }));
+            dispatchEvent(ErrorEvent::create(eventNames().navigateerrorEvent, errorMessage, errorInformation.sourceURL, errorInformation.line, errorInformation.column, { protectedScriptExecutionContext()->globalObject()->vm(), result }));
```
Do we want to do this if there is no error information?

rwlbuis replied:
Ditto.

rwlbuis committed:
[Navigation] Provide better error reporting
https://bugs.webkit.org/show_bug.cgi?id=279032

Reviewed by Alex Christensen.

Provide better error reporting by refactoring code in SerializedScriptValue that extracts detailed
error information from JSC::ErrorInstance and reuse the new method both for error reporting when
aborting the navigation and on handler promise rejection.

* LayoutTests/http/wpt/navigation-api/transition-promises-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/signal-abort-window-stop-after-intercept-expected.txt:
* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
(WebCore::CloneSerializer::dumpIfTerminal):
(WebCore::extractErrorInformationFromErrorInstance):
* Source/WebCore/bindings/js/SerializedScriptValue.h:
* Source/WebCore/page/Navigation.cpp:
(WebCore::Navigation::abortOngoingNavigation):
(WebCore::Navigation::innerDispatchNavigateEvent):

Canonical link: https://commits.webkit.org/283197@main
